### PR TITLE
[transformer] Allow for different backend for Pipeline Parallel ProcessGroups

### PR DESCRIPTION
**Pull request by crcrpar**
This PR enables us to specify `torch.distributed.Backend` for PipelineParallel Process Groups and the other PGs by adding keyword arguments of `"default_backend"` and `"p2p_backend"` to `apex.transformer.parallel_state.initialize_model_parallel`.

By running `tests/L0/run_transformer/run_gpt_minimal_test.py` with various tensor|pipeline parallel size on a node of 8 GPUs, I've verified UCC for Pipeline Parallel PGs and NCCL for any others is functional. NCCL for PP PGs and UCC for the other PGs is not verified and may have bugs, which is why these warnings exist [here](https://github.com/NVIDIA/apex/compare/master...crcrpar:pipelineparallel_ucc_backend?expand=1#diff-bb2dfca8484e986cfb3277f966640c6af6f214aa6f3c5cb713994c72c67f7addR119-R126).

The main changes are in `apex/transformer/parallel_state.py` and most of changes are in testing code/unittests.

## Known limitations
- NCCL for Pipeline Parallel PG & UCC for the others don't seem to work
- `_forward_backward_pipelining_with_interleaving` hasn't been tested well with hybrid backends

## Development environment
- nvcr.io/nvidia/pytorch:22.04-py3 with UCC v1.0.0 and torch_ucc of 426de4a0eb9059760967c06491598ec194150395
- host with a driver >= 470.42.01 due to https://github.com/NVIDIA/nvidia-settings/blame/5b455b89bb73f56818c84444806bc9c928da67ac/src/nvml.h#L6009-L6026. Even if you'll use a container the driver's version seems to matter in my experience.

## TODOs
Okay to access in follow-up PRs.
- Improve docstring
- Figure out limitations

cc @eqy @ptrblck @csarofeen @zasdfgbnm @Aidyn-A

### CODE REVIEW

File: apex/transformer/testing/distributed_test_base.py [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
 
-    BACKEND_NCCL = "nccl"
+class DistributedTestBase(common_distributed.MultiProcessTestCase):
 
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
     def setUp(self) -> None:
         super().setUp()
+        self._setup_pre_spawn()
```
Comment (eqy) [inline apex/transformer/testing/distributed_test_base.py:25]: Is this needed if the method is currently just `pass`?

Comment (crcrpar) [reply apex/transformer/testing/distributed_test_base.py:25]: `UccDistributedTestBase` uses this see around L.88

Comment (eqy) [reply apex/transformer/testing/distributed_test_base.py:25]: Thanks, my bad, thought this was for the NCCL test subclass

File: tests/L0/run_transformer/run_gpt_minimal_test.py [hunk 2]
```diff
@@ -1,194 +1,205 @@
...
+import time
+
+import torch
+try:
+    import torch_ucc
+except ImportError:
+    HAS_TORCH_UCC = False
+else:
+    HAS_TORCH_UCC = True
+    print("Use UCC as backend of Pipeline Parallel ProcessGroups")
```
Comment (crcrpar) [inline tests/L0/run_transformer/run_gpt_minimal_test.py:12]: semantic diff 1

File: tests/L0/run_transformer/run_gpt_minimal_test.py [hunk 3]
```diff
@@ -1,194 +1,205 @@
...
+            )
+            world_size = torch.distributed.get_world_size()
+
+        print(args.tensor_model_parallel_size, "MODEL PARALLEL SIZE")
+
+        parallel_state.initialize_model_parallel(
+            tensor_model_parallel_size_=args.tensor_model_parallel_size,
+            pipeline_model_parallel_size_=args.pipeline_model_parallel_size,
+            default_backend="nccl",
+            p2p_backend="ucc" if HAS_TORCH_UCC else "nccl",
```
Comment (crcrpar) [inline tests/L0/run_transformer/run_gpt_minimal_test.py:183]: semantic diff 2

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
 from typing import Optional
 
 import torch
 from torch.testing._internal import common_utils
+try:
+    import torch_ucc
+except ImportError:
+    HAS_TORCH_UCC = False
+else:
+    HAS_TORCH_UCC = True
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:13]: this looks kind of redundant to me and I feel like getting this from `distributed_test_base.py`

**Comment by eqy** [comment]
`run_transformer` seems OK on A100

File: apex/transformer/testing/distributed_test_base.py [hunk 5]
```diff
@@ -58,3 +67,50 @@ def _run(cls, rank, test_name, file_name, pipe):
...
+
+class NcclDistributedTestBase(DistributedTestBase):
+
+    DISTRIBUTED_BACKEND = "nccl"
+
+
+@unittest.skipUnless(
+    HAS_TORCH_UCC,
+    "Requires [`torch_ucc`](https://github.com/facebookresearch/torch_ucc)",
+)
```
Comment (crcrpar) [inline apex/transformer/testing/distributed_test_base.py:83]: Ideally I want to check driver version as UCC requires >=470.42.01

Comment (crcrpar) [reply apex/transformer/testing/distributed_test_base.py:91]: resolved in https://github.com/NVIDIA/apex/pull/1380/commits/813bc451b8384f847f9b4ed27e4884c2bdc62b94

**Pull request merged**